### ISSUE TITLE
Rewards for bridge validators

### ISSUE THREAD

**Issue by akolotov** [poanetwork/poa-bridge-contracts#119]
First of all for security reason the validators must have two accounts (two private keys): one account to sign bridge transactions and another account to receieve fees.

That's why the contract deployment configuration must be extended as follows:
* the parameter `VALIDATORS_REWARD_ACCOUNTS`. It must have the same number of addresses as the parameter `VALIDATORS`. The first account from `VALIDATORS` is linked with the first account in `VALIDATORS_REWARD_ACCOUNTS`, the second account from `VALIDATORS` is linked with the second account in `VALIDATORS_REWARD_ACCOUNTS` and so.

It means that the validator contract must be extended to:
* allow getting the list of accounts for getting reward
* allow to add both accounts at once when a new validator is added
* allow to remove the reward account address as soon as the validator is removed

There are three possible approaches to proceed with covering the validators expences on the bridge:
- no fees at all. The validators will be subsidizied by sponsor. This is implemented in the current version of the bridge contracts.
- fixed fee. The amount of tips will be fixed and does not depend on the value of assets transferred through the bridge.
- relative fee. The defrayal depepends on the value of assets and can be expressed in percentages.

Another important note is that the way how fees are distributed among the validators could also be different: the same amount can be sent to every actual bridge validator or the value could depend on some stake that the validator used to be approved as the bridge validator.

It is assumed that the mode could be changed during the bridge contracts life time.

That's why the suggestion is to move all logic that implements the way to calculate and distribute the bridge fee in a separate contract and register its address in the bridge contracts. The methods of the contract responsible for fees will be accessible through `delegatecall` functionality so storage data available for the bridge contracts will be also available for the fee contract as well. Only the upgradability admin will be allowed to change the address of the fee contract. 

This will also allow
* to choose one of several upproaches on the bridge contracts deployment stage. 
* to upgrade the process how to work with fees without necessity to upgrade entire bridge contract.
* to implemnet different approaches one by one: as soon as one implemented it could be applied to the real contracts without need to wait for other approaches implementation.
* to reduce the size of the bridge contract.

It is suggested to consider three main rules for fees:
1. The value of transfer that has been confirmed (signed) by the validator must be delivered to the another side of the bridge without changes.
2. Fee must be hold only when enough number of validators confirmed the transfer.
3. It is not mandatory to write off fee on another side of the bridge only.

The fee contract will have several interfaces to be used by the bridge contracts:

```solidity
function calculateFee(uint256 _value, bool _recover) external view returns(uint256 fee);
```
returns fee that should be held for the transfer of the particular value. If _recover is `true`, amount of fee will be recovered from the value assuming that it was previously reduced by fee.  

```solidity
function distributeFeeFromSignatures(uint256 _fee) external;
function distributeFeeFromAffirmation(uint256 _fee) external;
```
distribute the corresponding amount of fee among the bridge validators. It will access the bridge validators' reward accounts and distribute the specified amount of fee as per the implemented approach. This method must be allowed for invocation only by the bridge contract.

```solidity
function setFee(uint256 _fee) external;
```
set new amount for fee. Depending of the type of fee: fixed or relative, the amount could either specify exact amount of tokens/coins to be used for fee or a fraction of some entire value. This method must be allowed for the invocation by the bridge admin account only.	

```solidity
function getFee() external view returns(uint256 fee);
```
returns the amount of fee.

The following example can clarify how these new methods will be used in `erc20-to-native` mode:

`ForeignBridgeErcToNative`:
  * will have the fee contract set as zero;

`HomeBridgeErcToNative`:
  * `function ()`:
  ```solidity
        IBlockReward blockReward = blockRewardContract();
        uint256 totalMinted = blockReward.mintedTotallyByBridge(address(this));
        require(msg.value <= totalMinted.sub(totalBurntCoins()));
        setTotalSpentPerDay(getCurrentDay(), totalSpentPerDay(getCurrentDay()).add(msg.value));
        uint256 valueToTransfer = msg.value;
        if (feeContract() != address(0)) {
        	uint256 fee = feeContract().delegatecall(abi.encodeWithSignature("calculateFee(uint256,bool)", valueToTransfer, false));
        	valueToTransfer = valueToTransfer - fee;
        }
        setTotalBurntCoins(totalBurntCoins().add(valueToTransfer));
        address(0).transfer(valueToTransfer);
        emit UserRequestForSignature(msg.sender, valueToTransfer);
  ```
  * `onExecuteAffirmation`:
  ```solidity
        IBlockReward blockReward = blockRewardContract();
        require(blockReward != address(0));
        uint256 valueToMint = msg.value;
        if (feeContract() != address(0)) {
        	uint256 fee = feeContract().delegatecall(abi.encodeWithSignature("calculateFee(uint256,bool)", valueToMint, false));
        	feeContract().delegatecall(abi.encodeWithSignature("distributeFeeFromAffirmation(uint256)", fee));
        	valueToMint = valueToMint - fee;
        }
        blockReward.addExtraReceiver(valueToMint, _recipient);
        return true;
  ```
`BasicHomeBridge`:
  * `submitSignature`:
  ```solidity
        uint256 reqSigs = requiredSignatures();
        if (signed >= reqSigs) {
            setNumMessagesSigned(hashMsg, markAsProcessed(signed));
            emit CollectedSignatures(msg.sender, hashMsg, reqSigs);


	        if (feeContract() != address(0)) {
		        address recipient;
		        uint256 amount;
		        bytes32 txHash;
		        address contractAddress;
		        (recipient, amount, txHash, contractAddress) = Message.parseMessage(message);
        		uint256 fee = feeContract().delegatecall(abi.encodeWithSignature("calculateFee(uint256,bool)", amount, true));
        		feeContract().delegatecall(abi.encodeWithSignature("distributeFeeFromSignatures(uint256)", fee));
		    }

        }
  ```

`feeContract`:
```solidity
function calculateFee(uint256 _value, bool _recover) external view returns(uint256 fee) {
	uint256 fee = getFee();
	if (!_recover) {
		return _value * fee / 1 ether;
	}
	return _value*fee/(1 ether - fee);
}
```
And for this case
* `distributeFeeFromSignatures` will call `blockReward.addExtraReceiver` for every reward validator address and the value will be equal `(fee / number of validators)`.
* `distributeFeeFromSignatures` will trnasfer coins to every reward validator address, the value of transfer will be `(fee / number of validators)`.

---

Parent of #122 
Parent of #123 
Parent of #124 
Parent of #125 
Parent of #126 
Parent of #127 
Parent of poanetwork/bridge-ui#163

**Comment by fernandomg** [comment]
@akolotov, a few questions that arise after reviewing with @patitonar:

1. About fee distribution, it should be distributed among all the validators or only among the ones that signed the transfer?

2. In regard to `reward` addresses, the idea is to keep them indexed in an `address` mapping, with the `validator address` as the key. With that in mind, we need to recover the list of validators to be able to obtain every reward address, so we can properly distribute the `fee`. Right now we don't see a way to recover the list of validators. Do you have any suggestion on how to approach this scenario?

**Comment by akolotov** [comment]
@fernandomg please find my answers below:

> About fee distribution, it should be distributed among all the validators or only among the ones that signed the transfer?

Among all the validators. The reason is because if we need just 1 of 3 signatures (for example) only one transaction will be accepted and 2 transactions will be reverted, so, the contract will know nothing that other 2 validators also actively participated in the transfer.  

> In regard to reward addresses, the idea is to keep them indexed in an address mapping, with the validator address as the key. With that in mind, we need to recover the list of validators to be able to obtain every reward address, so we can properly distribute the fee. Right now we don't see a way to recover the list of validators. Do you have any suggestion on how to approach this scenario?

How about the following idea:
* one map to keep relationship between the validator address and the corresponding reward address (as you already mentioned): `Reward[V] = R`, where `V` - a validator address, `R` - a reward address;
* two maps to build doubly linked list: the first map to keep the next element in the list, the second map to keep the previous element in the list.
![formula](https://latex.codecogs.com/gif.latex?\inline&space;Next[address(0)]&space;=&space;V_1;&space;Next[V_i]&space;=&space;V_{i&plus;1};&space;Next[V_n]&space;=&space;address(0))
![formula](https://latex.codecogs.com/gif.latex?\inline&space;Prev[address(0)]&space;=&space;V_n;&space;Prev[V_i]&space;=&space;V_{i-1};&space;Prev[V_1]&space;=&space;address(0))

**Comment by patitonar** [comment]
@akolotov here are some questions:
> There are three possible approaches to proceed with covering the validators expences on the bridge:
no fees at all. The validators will be subsidizied by sponsor. This is implemented in the current version of the bridge contracts.
fixed fee. The amount of tips will be fixed and does not depend on the value of assets transferred through the bridge.
relative fee. The defrayal depepends on the value of assets and can be expressed in percentages.

1. In order to determine if the bridge uses `fixed fee` or `relative fee` should we add the ability to set a type of fee parameter on the bridge contract?

> Another important note is that the way how fees are distributed among the validators could also be different: the same amount can be sent to every actual bridge validator or the value could depend on some stake that the validator used to be approved as the bridge validator.

2. Similar to question `1`, should we add the ability to set a distribution mode parameter on the bridge contract?

3. Can you add more details on how distribution by stake should work?

> uint256 fee = feeContract().delegatecall(abi.encodeWithSignature("calculateFee(uint256,bool)", amount, true));

4. That `delegatecall` method call to `calculateFee` is used on some of the examples. The version of solidity used on the contracts is `0.4.24`
In [this version](https://solidity.readthedocs.io/en/v0.4.24/units-and-global-variables.html?highlight=delegatecall#address-related), `delegatecall` is defined as:
```
<address>.delegatecall(...) returns (bool)
```
From [version `0.5.0`](https://solidity.readthedocs.io/en/v0.5.2/units-and-global-variables.html#members-of-address-types), `delegatecall` is defined as:
```
<address>.delegatecall(bytes memory) returns (bool, bytes memory)
```
So using `delegatecall` to get the returned value for `calculateFee` would be possible only in this last version. Should we consider migrating the contracts implementation to `0.5.x` version? or should we think on another approach for calling `calculateFee` ?

**Comment by akolotov** [comment]
@patitonar 

For 1 and 2 the idea is to have a separate contract - the fee contract. Depending on the scheme of the fee calculation and distribution it could have different implementations. So, if we need to change the approach how to fee calculated we deploy another contract and register it in the bridge contracts.

In this particular PR we need to implement only one type of the fee contract: when the fee is distributed proportionally between the bridge validators. So, we don't need to worry how distribution by stake should work at this moment.

For the item 4, you can take a look how this functionality is already implemented in the Proxy contract:
https://github.com/poanetwork/poa-bridge-contracts/blob/46151629adb3dd20e6fb47d4ba91b88c6d388c4f/contracts/upgradeability/Proxy.sol#L59-L80

**Comment by akolotov** [comment]
Available in [the release 2.3.0](https://github.com/poanetwork/poa-bridge-contracts/releases/tag/2.3.0)
